src/jit/gcencode.cpp

```diff
@@ -106,6 +106,293 @@ ReturnKind GCInfo::getReturnKind()
+                        // lifetime for the part inside the filter and adjust
+                        // the start of the original lifetime to be the end
+                        // of the filter
+                        CLANG_FORMAT_COMMENT_ANCHOR;
+#ifdef DEBUG
+                        if (compiler->verbose)
+                        {
+                            printf("Splitting lifetime for filter.\nOld: ");
+                            gcDumpVarPtrDsc(varTmp);
+                        }
+#endif // DEBUG
+
+                        varPtrDsc* desc = new (compiler, CMK_GC) varPtrDsc;
+#ifndef JIT32_GCENCODER
+                        desc->vpdVarNum = varTmp->vpdVarNum | pinned_OFFSET_FLAG;
+                        desc->vpdBegOfs = begOffs;
+                        desc->vpdEndOfs = filterEnd;
+                        desc->vpdNext   = gcVarPtrList;
+                        gcVarPtrList    = desc;
+
+                        varTmp->vpdBegOfs = filterEnd;
+#else
+                        // Mark varTmp as pinned and generated use varPtrDsc(desc) as non-pinned
+                        // since gcInsertVarPtrDscSplit requires that varTmp->vpdBegOfs must precede desc->vpdBegOfs
+                        desc->vpdVarNum = varTmp->vpdVarNum;
+                        desc->vpdBegOfs = filterEnd;
+                        desc->vpdEndOfs = endOffs;
+                        gcInsertVarPtrDscSplit(desc, varTmp);
+
+                        varTmp->vpdVarNum = varTmp->vpdVarNum | pinned_OFFSET_FLAG;
+                        varTmp->vpdEndOfs = filterEnd;
+#endif
+
+#ifdef DEBUG
+                        if (compiler->verbose)
+                        {
+                            printf("New (1 of 2): ");
+                            gcDumpVarPtrDsc(desc);
+                            printf("New (2 of 2): ");
+                            gcDumpVarPtrDsc(varTmp);
+                        }
+#endif // DEBUG
+                    }
+                    else
+                    {
+                        // The variable lifetime is completely within the filter,
+                        // so just add the pinned flag.
+                        CLANG_FORMAT_COMMENT_ANCHOR;
+#ifdef DEBUG
+                        if (compiler->verbose)
+                        {
+                            printf("Pinning lifetime for filter.\nOld: ");
+                            gcDumpVarPtrDsc(varTmp);
+                        }
+#endif // DEBUG
+
+                        varTmp->vpdVarNum |= pinned_OFFSET_FLAG;
+#ifdef DEBUG
+                        if (compiler->verbose)
+                        {
+                            printf("New : ");
+                            gcDumpVarPtrDsc(varTmp);
+                        }
+#endif // DEBUG
+                    }
+                }
+            }
+        } // HasFilter
+    }     // Foreach EH
+}
+
+void GCInfo::gcInsertVarPtrDscSplit(varPtrDsc* desc, varPtrDsc* begin)
+{
```

Comment:
@BruceForstall 

> why do we need to sort the insertions for JIT32_GCENCODER but not for the other case?

To summarize,
- x86/Windows : We don't need `gcMarkFilterVarsPinned` and `gcInsertVarPtrDscSplit`.
- x86/Linux : We do need `gcMarkFilterVarsPinned` but we have **x86-GCInfo**. It requires varPtrList to be sorted by vpdBegOfs ascending. That's why we need `gcInsertVarPtrDscSplit` for only this case.
- the others : We do need `gcMarkFilterVarsPinned` and **GCInfoEncoder** does not require varPtrList is sorted.

> Could this function be made to work for both (e.g., #ifdef to sort for JIT32_GCENCODER and not worry about sorting for the opposite?) That would reduce #ifdefs in GCInfo::gcMarkFilterVarsPinned().

Thank you for pointing that out. When I first wrote this patch, I just did not want to change the current behavior just in case. I'll try what you said then apply it to the patch.